### DISCUSSION TITLE
solver did not converge - serpentization modeling

### DISCUSSION THREAD

**Discussion by erexshah** [reaktoro/reaktoro#403] [Q&A]
### Description

Hello Allan, 

I'm struggling with the equilibrium calculation of a mineral/water/gas system. When I include all mineral phases, the solver does not converge. However, if some minerals are excluded (simplifying the system), the calculation succeeds. Could you please check whether the issue is with the inputs or the solver algorithm?

Many thanks, 
Arash

### Example

```py
import reaktoro as rkt

# Weight of primary minerals in the rock
weight_olivine:float = 800.0 # grams
weight_orthopyroxene:float = 150.0 # grams
weight_clinopyroxene:float = 50.0 # grams

weight_fluid:float = 1000.0 # grams

# Primary Minerals Composition
mol_frac_Forsterite:float = 0.90 # Olivine Fo_90
mol_frac_Enstatite:float = 0.85 #Orthopyroxene En_85
mol_frac_Diopside:float = 0.90 #Clinopyroxene Di_90

db:rkt.SupcrtDatabase = rkt.SupcrtDatabase("supcrt98")

solution_species:list[str] = [
    "H2O(aq)",
    "H+",
    "Na+",
    "Cl-",
    "HCO3-",
    "Ca+2",
    "Mg+2",
    "K+",
    "SiO2(aq)",
    "Fe+2",
    "SO4-2",
    "O2(aq)",
    "H2(aq)"
]

solution:rkt.AqueousPhase = rkt.AqueousPhase(solution_species)
solution.set(rkt.ActivityModelDebyeHuckel())

clinopyroxene:rkt.MineralPhase = rkt.MineralPhase("Diopside Hedenbergite")
clinopyroxene.setName("Clinopyroxene")

orthopyroxene:rkt.MineralPhase = rkt.MineralPhase("Enstatite Ferrosilite")
orthopyroxene.setName("Orthopyroxene")

olivine:rkt.MineralPhase = rkt.MineralPhase("Forsterite Fayalite")
olivine.setName("Olivine")

magnetite:rkt.MineralPhase = rkt.MineralPhase("Magnetite")

brucite:rkt.MineralPhase = rkt.MineralPhase("Brucite")

serpentine:rkt.MineralPhase = rkt.MineralPhase("Chrysotile Greenalite")
serpentine.setName("Serpentine")

talc:rkt.MineralPhase = rkt.MineralPhase("Talc Minnesotaite")
talc.setName("Talc")

amphibole:rkt.MineralPhase = rkt.MineralPhase("Tremolite Ferrotremolite")
amphibole.setName("Amphibole")

gases:rkt.MineralPhase = rkt.GaseousPhase("H2(g) O2(g) H2O(g)")
gases.set(rkt.ActivityModelPengRobinson())

system:rkt.ChemicalSystem = rkt.ChemicalSystem(db, solution, gases,
                                               clinopyroxene, orthopyroxene,
                                               olivine, magnetite,
                                               brucite, serpentine, talc,
                                               amphibole)

state:rkt.ChemicalState = rkt.ChemicalState(system)
state.setTemperature(200, "celsius")
state.setPressure(35, "MPa")

state.set("H2O(aq)", 1.0, "kg")
state.set("H+", 10**(-7.8), "mol")
state.set("Na+", 464.0, "mmol")
state.set("Cl-", 546, "mmol")
state.set("HCO3-", 2.3, "mmol")
state.set("Ca+2", 10.2, "mmol")
state.set("Mg+2", 24.8, "mmol")
state.set("K+", 9.8, "mmol")
state.set("SiO2(aq)", 0.16, "mmol")
state.set("Fe+2", 1.5e-6, "mmol")
state.set("SO4-2", 1e-15, "mmol")
state.set("O2(aq)", 1e-15, "mmol")
state.set("H2(aq)", 1e-15, "mmol")
state.set("H2(g)", 1e-15, "mmol")
state.set("H2O(g)", 1e-15, "mmol")
state.set("O2(g)", 1e-15, "mmol")

state.set("Diopside", mol_frac_Diopside, "mol")
state.set("Hedenbergite", 1.0-mol_frac_Diopside, "mol")

state.set("Enstatite",  mol_frac_Enstatite, "mol")
state.set("Ferrosilite", 1.0-mol_frac_Enstatite, "mol")

state.set("Forsterite", mol_frac_Forsterite, "mol")
state.set("Fayalite",  1.0-mol_frac_Forsterite, "mol")

state.scalePhaseMass("Orthopyroxene", weight_orthopyroxene, "g")
state.scalePhaseMass("Clinopyroxene", weight_clinopyroxene, "g")
state.scalePhaseMass("Olivine", weight_olivine, "g")
state.scalePhaseMass("AqueousPhase", weight_fluid, "g")

print(state)

solver:rkt.EquilibriumSolver = rkt.EquilibriumSolver(system)
result:rkt.EquilibriumResult = solver.solve(state)

print("Succussful?", result.succeeded())

print(state)

```

## The following is the simpler case where the solver successfully converged.

```py
import reaktoro as rkt

# Weight of primary minerals in the rock
weight_olivine:float = 1000.0 # grams
# weight_orthopyroxene:float = 150.0 # grams
# weight_clinopyroxene:float = 50.0 # grams

weight_fluid:float = 1000.0 # grams

# Primary Minerals Composition
mol_frac_Forsterite:float = 0.90 # Olivine Fo_90
# mol_frac_Enstatite:float = 0.85 #Orthopyroxene En_85
# mol_frac_Diopside:float = 0.90 #Clinopyroxene Di_90

db:rkt.SupcrtDatabase = rkt.SupcrtDatabase("supcrt98")

solution_species:list[str] = [
    "H2O(aq)",
    "H+",
    "Na+",
    "Cl-",
    "HCO3-",
    "Ca+2",
    "Mg+2",
    "K+",
    "SiO2(aq)",
    "Fe+2",
    "SO4-2",
    "O2(aq)",
    "H2(aq)"
]

solution:rkt.AqueousPhase = rkt.AqueousPhase(solution_species)
solution.set(rkt.ActivityModelDebyeHuckel())

# clinopyroxene:rkt.MineralPhase = rkt.MineralPhase("Diopside Hedenbergite")
# clinopyroxene.setName("Clinopyroxene")

# orthopyroxene:rkt.MineralPhase = rkt.MineralPhase("Enstatite Ferrosilite")
# orthopyroxene.setName("Orthopyroxene")

olivine:rkt.MineralPhase = rkt.MineralPhase("Forsterite Fayalite")
olivine.setName("Olivine")

magnetite:rkt.MineralPhase = rkt.MineralPhase("Magnetite")

brucite:rkt.MineralPhase = rkt.MineralPhase("Brucite")

serpentine:rkt.MineralPhase = rkt.MineralPhase("Chrysotile Greenalite")
serpentine.setName("Serpentine")

# talc:rkt.MineralPhase = rkt.MineralPhase("Talc Minnesotaite")
# talc.setName("Talc")

# amphibole:rkt.MineralPhase = rkt.MineralPhase("Tremolite Ferrotremolite")
# amphibole.setName("Amphibole")

gases:rkt.MineralPhase = rkt.GaseousPhase("H2(g) O2(g) H2O(g)")
gases.set(rkt.ActivityModelPengRobinson())

system:rkt.ChemicalSystem = rkt.ChemicalSystem(db, solution, gases,
                                               # clinopyroxene, orthopyroxene,
                                               olivine, magnetite,
                                               brucite, serpentine)#, talc,
                                               # amphibole)

state:rkt.ChemicalState = rkt.ChemicalState(system)
state.setTemperature(200, "celsius")
state.setPressure(35, "MPa")

state.set("H2O(aq)", 1.0, "kg")
state.set("H+", 10**(-7.8), "mol")
state.set("Na+", 464.0, "mmol")
state.set("Cl-", 546, "mmol")
state.set("HCO3-", 2.3, "mmol")
state.set("Ca+2", 10.2, "mmol")
state.set("Mg+2", 24.8, "mmol")
state.set("K+", 9.8, "mmol")
state.set("SiO2(aq)", 0.16, "mmol")
state.set("Fe+2", 1.5e-6, "mmol")
state.set("SO4-2", 1e-15, "mmol")
state.set("O2(aq)", 1e-15, "mmol")
state.set("H2(aq)", 1e-15, "mmol")
state.set("H2(g)", 1e-15, "mmol")
state.set("H2O(g)", 1e-15, "mmol")
state.set("O2(g)", 1e-15, "mmol")

# state.set("Diopside", mol_frac_Diopside, "mol")
# state.set("Hedenbergite", 1.0-mol_frac_Diopside, "mol")

# state.set("Enstatite",  mol_frac_Enstatite, "mol")
# state.set("Ferrosilite", 1.0-mol_frac_Enstatite, "mol")

state.set("Forsterite", mol_frac_Forsterite, "mol")
state.set("Fayalite",  1.0-mol_frac_Forsterite, "mol")

# state.scalePhaseMass("Orthopyroxene", weight_orthopyroxene, "g")
# state.scalePhaseMass("Clinopyroxene", weight_clinopyroxene, "g")
state.scalePhaseMass("Olivine", weight_olivine, "g")
state.scalePhaseMass("AqueousPhase", weight_fluid, "g")

print(state)

solver:rkt.EquilibriumSolver = rkt.EquilibriumSolver(system)
result:rkt.EquilibriumResult = solver.solve(state)

print("Succussful?", result.succeeded())

print(state)
```

### What is your operating system?

Linux

**Comment by allanleal** [comment]
Hi Arash, thanks for opening this discussion. It seems your chemical system and its initial state tends to consume all H2O(aq). I removed the gas phase from the system definition to ensure H2O(aq) was not being transformed into H2O(g). 

If we add the restriction below, your original more complicated example works:

```py
restrictions = rkt.EquilibriumRestrictions(system)
restrictions.cannotReact("H2O(aq)")

result:rkt.EquilibriumResult = solver.solve(state, restrictions)
```

But I think this numerical restriction is not adequate. It forces H2O(aq) to maintain its initial amount; thus, H2O(aq) cannot be formed nor dissociated.

I noticed that the solid solutions **Talc** (end-members: Talc and Minnesotaite) and **Amphibole** (end-members: Tremolite and Ferrotremolite) are causing trouble. Removing them makes the calculation succeed.

Do you have any insights here that could help? Experimental data indicating what should be the equilibrium state would also be helpful, but I suspect this is not available. Trying a more recent database that perhaps have revised/calibrated thermodynamic parameters for these minerals could also possibly help.

**Comment by erexshah** [comment]
Hi Allan,

Thank you so much for your quick reply.

Water is one of the main reactants in this system (primary minerals + water <--> secondary minerals + hydrogen), so we cannot exclude it from the reaction. However, we can remove the gas phase at this step.

I'm trying to reproduce the results published [here](https://doi.org/10.1016/j.gca.2008.10.032). The authors used SUPCRT92 for the database and EQ3/6 for the equilibrium calculations. Therefore, I need to use the same database. As you pointed out, there might be issues with the thermodynamic data, so I checked the supcrt98.yaml file and noticed that the Gibbs energy for Minnesotaite was set to 0, which I have now corrected. This could be one of the reasons the solver failed when Amphibole was included in the system. Nonetheless, the solver still fails. I may need to further customize the database, though I hope the issue is not with the optimization algorithm (GEM vs. LMA).

I need to ensure I can reproduce the same results using Reaktoro before proceeding to the next steps.

Regarding the chemical system, these reactions are well-studied, and the amount of water is more than sufficient. There are dozens of equilibrium calculations for this system under different conditions [here](https://doi.org/10.1016/j.lithos.2013.03.008) and [here](https://doi.org/10.1029/2022GC010658).

**Comment by allanleal** [comment]
Hi Arash,

The system of non-linear equations used in Reaktoro for solving chemical equilibrium in a framework for Gibbs energy minimization (GEM) can be cast to law of mass action (LMA) equations. GEM and LMA are essentially solving the same mathematical problem, but the GEM formulation has, in my opinion, a superior structure for handling many phases and finding their stability. More on this paper:

_Leal, A., Kulik, D., Smith, W. & Saar, M. (2017). An overview of computational methods for chemical equilibrium and kinetic calculations for geochemical and reactive transport modeling. Pure and Applied Chemistry, 89(5), 597-643. https://doi.org/10.1515/pac-2016-1107_

So, GEM and LMA should not be the issue here, but the thermodynamic properties. In the [original SUPCRT92 database slop98.dat](https://github.com/reaktoro/reaktoro/blob/main/utilities/supcrt-parser/original.slop98.dat), this is what we find:

```text
 MINNESOTAITE        Fe3Si4O10(OH)2                
 Mnn                 Fe(3)Si(4)O(12)H(2)                     
 ref:1               5.May.78 
          999999.0      999999.0    83.500   147.860
         88.310000     42.610000    -11.150000
         800.00
```

The standard-state Gibbs energy and enthalpy of formation for this mineral are set to 999999.0, which means unknown. Looks like I changed this to zero when porting to YAML and JSON formats. Should in fact have set to null. Note that a very high positive G° value for a substance will cause it to never appear in the equilibrium state (could also cause numerical issues). 

I see in the paper https://linkinghub.elsevier.com/retrieve/pii/S0016703708006467 that Fe(OH)2 was considered (Table 3) and values for $\Delta G_f^\circ$ given for Minnesotaite. Fe(OH)2 is not present in the supcrt98 database. Have you included it? Also note that the authors considered two different values of $\Delta G_f^\circ$ for Fe(OH)2; the different modeling outcomes shown in Figure 8. 

I've noticed that the database core10.dat supported in Reaktoro also contains these minerals. Maybe worth considering it. 

I'd be happy to work on this with you; would you be able to share via email the most recent files you have (adapted YAML database and python code)? I'll take a look.

**Reply by erexshah** [reply]
Hi Allan,

Thank you for the explanation. I agree, the issue is most likely related to the database. I really appreciate your offer to help and will reach out via email to share the files.

Could we mark this discussion as resolved for now?

**Reply by allanleal** [reply]
Let's do that! Thanks.
